Riot/Modules/VoiceBroadcast/VoiceBroadcastServiceProvider.swift

```diff
@@ -0,0 +1,107 @@
+class VoiceBroadcastServiceProvider {
+    
+    // MARK: - Constants
+    
+    static let shared = VoiceBroadcastServiceProvider()
+    
+    // MARK: - Properties
+    
+    // VoiceBroadcastService in the current session
+    public var voiceBroadcastService: VoiceBroadcastService? = nil
+    
+    // MARK: - Setup
+    
+    private init() {}
+    
+    // MARK: - Public
+    
+    public func getOrCreateVoiceBroadcastService(for room: MXRoom, completion: @escaping (VoiceBroadcastService?) -> Void) {
+        guard let voiceBroadcastService = self.voiceBroadcastService else {
+            self.setupVoiceBroadcastService(for: room) { voiceBroadcastService in
+                completion(voiceBroadcastService)
+            }
+            return
+        }
+        
+        if voiceBroadcastService.room.roomId == room.roomId {
+            completion(voiceBroadcastService)
+        }
+        
+        completion(nil)
+    }
+    
+    public func tearDownVoiceBroadcastService() {
+                
+        self.voiceBroadcastService = nil
+
+        MXLog.debug("Stop monitoring voice broadcast recording")
+    }
+    
+    // MARK: - Private
+    
+    // MARK: VoiceBroadcastService setup
+    
+    private func createVoiceBroadcastService(for room: MXRoom, state: VoiceBroadcastService.State) {
+                
+        let voiceBroadcastService = VoiceBroadcastService(room: room, state: VoiceBroadcastService.State.stopped)
+        
+        self.voiceBroadcastService = voiceBroadcastService
+        
+        MXLog.debug("Start monitoring voice broadcast recording")
+    }
+    
+    private func setupVoiceBroadcastService(for room: MXRoom, completion: @escaping (VoiceBroadcastService?) -> Void) {
+        self.getLastVoiceBroadcastInfo(for: room) { event in
+            guard let voiceBroadcastInfoEvent = event else {
+                self.createVoiceBroadcastService(for: room, state: VoiceBroadcastService.State.stopped)
+                completion(self.voiceBroadcastService)
+                return
+            }
+            
+            guard let voiceBroadcastInfoEventContent = VoiceBroadcastEventContent(fromJSON: voiceBroadcastInfoEvent.content) else {
+                self.createVoiceBroadcastService(for: room, state: VoiceBroadcastService.State.stopped)
+                completion(self.voiceBroadcastService)
+                return
+            }
+            
+            if voiceBroadcastInfoEventContent.state == VoiceBroadcastService.State.stopped.rawValue {
+                self.createVoiceBroadcastService(for: room, state: VoiceBroadcastService.State.stopped)
+                completion(self.voiceBroadcastService)
+            } else if voiceBroadcastInfoEvent.stateKey == room.mxSession.myUserId {
+                self.createVoiceBroadcastService(for: room, state: VoiceBroadcastService.State(rawValue: voiceBroadcastInfoEventContent.state) ?? VoiceBroadcastService.State.stopped)
```

Comment:
This use case is particular: the end user started a voice broadcast, but no service is running locally. This may be due the following situations:
- the application crashed
- the voice broadcast has been started from another session

We may block the end user here by returning nil here
or we may let him get back the control of his voice broadcast

Reply:
I documented the function.